grub-core/kern/efi/tpm.c

```diff
@@ -0,0 +1,251 @@
+
+    switch (status) {
+    case GRUB_EFI_SUCCESS:
+      return 0;
+    case GRUB_EFI_DEVICE_ERROR:
+      return grub_error (GRUB_ERR_IO, N_("Command failed"));
+    case GRUB_EFI_INVALID_PARAMETER:
+      return grub_error (GRUB_ERR_BAD_ARGUMENT, N_("Invalid parameter"));
+    case GRUB_EFI_BUFFER_TOO_SMALL:
+      return grub_error (GRUB_ERR_BAD_ARGUMENT, N_("Output buffer too small"));
+    case GRUB_EFI_NOT_FOUND:
+      return grub_error (GRUB_ERR_UNKNOWN_DEVICE, N_("TPM unavailable"));
+    default:
+      return grub_error (GRUB_ERR_UNKNOWN_DEVICE, N_("Unknown TPM error"));
+    }
+  } else {
+    tpm2 = grub_efi_open_protocol (tpm_handle, &tpm2_guid,
+				   GRUB_EFI_OPEN_PROTOCOL_GET_PROTOCOL);
+
+    if (!grub_tpm2_present(tpm2)) {
+      return 0;
+    }
+
+    /* UEFI TPM protocol takes the raw operand block, no param block header */
+    status = efi_call_5 (tpm2->submit_command, tpm2,
+			 inbuf->IPBLength - inhdrsize, inbuf->TPMOperandIn,
+			 outbuf->OPBLength - outhdrsize, outbuf->TPMOperandOut);
+
+    switch (status) {
+    case GRUB_EFI_SUCCESS:
+      return 0;
+    case GRUB_EFI_DEVICE_ERROR:
+      return grub_error (GRUB_ERR_IO, N_("Command failed"));
+    case GRUB_EFI_INVALID_PARAMETER:
+      return grub_error (GRUB_ERR_BAD_ARGUMENT, N_("Invalid parameter"));
+    case GRUB_EFI_BUFFER_TOO_SMALL:
+      return grub_error (GRUB_ERR_BAD_ARGUMENT, N_("Output buffer too small"));
+    case GRUB_EFI_NOT_FOUND:
+      return grub_error (GRUB_ERR_UNKNOWN_DEVICE, N_("TPM unavailable"));
+    default:
+      return grub_error (GRUB_ERR_UNKNOWN_DEVICE, N_("Unknown TPM error"));
+    }
+  }
+}
+
+typedef struct {
+	grub_uint32_t pcrindex;
+	grub_uint32_t eventtype;
+	grub_uint8_t digest[20];
+	grub_uint32_t eventsize;
+	grub_uint8_t event[1];
+} Event;
+
+grub_err_t
+grub_tpm_log_event(unsigned char *buf, grub_size_t size, grub_uint8_t pcr,
+		   const char *description)
+{
+  grub_efi_handle_t tpm_handle;
+  grub_efi_status_t status;
+  grub_efi_tpm_protocol_t *tpm;
+  grub_efi_tpm2_protocol_t *tpm2;
+  grub_efi_physical_address_t lastevent;
+  grub_efi_uint8_t protocol_version;
+  grub_uint32_t eventnum = 0;
+  grub_uint32_t algorithm;
+  Event *event;
+  EFI_TCG2_EVENT *event2;
+
+  if (!grub_tpm_handle_find(&tpm_handle, &protocol_version)) {
+    return 0;
+  }
+
+  if (protocol_version == 1) {
+    tpm = grub_efi_open_protocol (tpm_handle, &tpm_guid,
+				  GRUB_EFI_OPEN_PROTOCOL_GET_PROTOCOL);
+
+    if (!grub_tpm_present(tpm)) {
+      return 0;
+    }
+
+    event = grub_zalloc(sizeof (Event) + grub_strlen(description) + 1);
+    if (!event) {
+      grub_printf("No buffer\n");
+      return grub_error (GRUB_ERR_OUT_OF_MEMORY,
+			 N_("cannot allocate TPM event buffer"));
+    }
+
+    event->pcrindex = pcr;
+    event->eventtype = 0x0d;
+    event->eventsize = grub_strlen(description) + 1;
+    grub_memcpy(event->event, description, event->eventsize);
+
+    algorithm = 0x00000004; /* SHA 1 */
```

Comment:
can we give these magic numbers symbolic names in the form of #defines or something? (the 0x0d above as well)